[WIP] Add a comprehensive API

tkurki opened the pull request:
This PR is for formulating a comprehensive JavaScript Signal K client with access to 
- connection
- stream API
- rest API
- schema
- discovery

Some of the methods may be available inside the Node server to provide a standard interface to the Signal K apis for plugins.

See also https://github.com/SignalK/specification/issues/244

tkurki committed:
Add new api prototype

tkurki committed:
Add & update

timmathews committed:
Add JSDoc documentation

timmathews committed:
Convert design.md to JSDoc on stubbed index.js

timmathews committed:
Write documentation to README.md

timmathews committed:
Latest versions of dependencies

timmathews committed:
Implement Constructor

Add constructor functionality with tests

timmathews committed:
Stub tests for connect function

timmathews committed:
Clean up package.json

timmathews committed:
Implement Client#connect

Client#connect and associated tests.

TODO: Create skConnection object with all properties

timmathews committed:
Consolidate approaches to new API

timmathews committed:
Update all deps to latest versions

timmathews commented:
Thought it might be worthwhile to combine the API work I was doing on the documentation branch with this branch

timmathews committed:
Implement subscribe and subscribeAll

subscribe and subscribeAll return a promise and handle WebSocket
connection. connect only establishes connection with the server, which
in the case of Signal K means only that makes an HTTP request to
/signalk and ensures that the server provides an endpoint for the
version of Signal K that the client is expecting.

All of the WebSocket setup happens in subscribe now. subscribeAll is
just a wrapper around subscribe for the most common use case;

timmathews committed:
Ignore *.log
